app/models/topic.rb

```diff
@@ -634,17 +634,37 @@ def self.similar_to(title, raw, user = nil)
 
     tsquery = Search.to_tsquery(term: tsquery, joiner: "|")
 
+    guardian = Guardian.new(user)
+
     candidates = Topic
       .visible
       .listable_topics
-      .secured(Guardian.new(user))
+      .secured(guardian)
       .joins("JOIN topic_search_data s ON topics.id = s.topic_id")
       .joins("LEFT JOIN categories c ON topics.id = c.topic_id")
       .where("search_data @@ #{tsquery}")
       .where("c.topic_id IS NULL")
       .order("ts_rank(search_data, #{tsquery}) DESC")
       .limit(SiteSetting.max_similar_results * 3)
 
+    excluded_category_ids = Set.new
+
+    excluded_category_ids.merge(
+      Category.secured(guardian).where(search_priority: Searchable::PRIORITIES[:ignore]).pluck(:id)
+    )
+
+    if user
+      excluded_category_ids.merge(
+        CategoryUser
+          .where(notification_level: CategoryUser.notification_levels[:muted], user: user)
+          .pluck(:category_id)
+      )
+    end
+
+    if excluded_category_ids.present?
+      candidates = candidates.where("topics.category_id NOT IN (?)", excluded_category_ids)
+    end
```

Comment:
hmm are you sure about that? 

category_id NOT in (SELECT ....) should work, worse case I guess you could use `WITH` which can force ordering for you? 

I worry about shuffling large lists of ids around for sites with tons of categories.

Reply:
> hmm are you sure about that?

Yes I'm quite sure joining would make the query perform significantly worst because we would have to join all rows in the `topics` table against the `categories` and `category_users` table as compared to just filtering against a set of ids on the `topics` table itself. 

> category_id NOT in (SELECT ....) should work, worse case I guess you could use WITH which can force ordering for you?

I've updated the query to use a subplan instead.

Reply:
looks good now! the interpolation can confuse and trigger various auto security scanners, I wonder if long term we have a nice internal pattern to say ... hey this interpolation is safe... nothing to change right now though.